### ISSUE TITLE
Non-fatal error while running a specific `cargo check` command

### ISSUE THREAD

**Issue by Kiwifuit** [rust-lang/rust#128130]
### Code
I'm sorry I cannot provide a minimal code example as I am equally confused with this issue. However, I have tried [this](https://files.catbox.moe/ngl98w.zip) and it produces the error.
The original repository can be found [here](https://github.com/Kiwifuit/McManager)

I am using rust-analyzer v0.3.2045 with the command it spat out:
```
cargo check --workspace --message-format=json-diagnostic-rendered-ansi --manifest-path <Cargo.toml> --keep-going --all-targets
```
No error is produced with `cargo check`, just with these specific flags

### Meta

`rustc --version --verbose`:
```
rustc 1.79.0 (129f3b996 2024-06-10)
binary: rustc
commit-hash: 129f3b9964af4d4a709d1383930ade12dfe7c081
commit-date: 2024-06-10
host: x86_64-unknown-linux-gnu
release: 1.79.0
LLVM version: 18.1.7
```
This issue only happens with the stable build

### Error output

```
thread 'rustc' panicked at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/compiler/rustc_query_system/src/dep_graph/serialized.rs:192:9:
thread 'rustc' panicked at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/compiler/rustc_query_system/src/dep_graph/serialized.rs:192:9:
assertion `left == right` failed
assertion `left == right` failed
  left: 3194662
  left: 3225253
 right: 0
 right: 0
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0: rust_begin_unwind
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panicking.rs:652:5
   1: core::panicking::panic_fmt
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/panicking.rs:72:14
   2: core::panicking::assert_failed_inner
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/panicking.rs:409:17
   3: core::panicking::assert_failed::<usize, usize>
   4: rustc_incremental::persist::load::setup_dep_graph
   5: <rustc_interface::queries::Queries>::global_ctxt
   6: rustc_interface::interface::run_compiler::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
   0: rust_begin_unwind
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panicking.rs:652:5

   1: core::panicking::panic_fmt
error: the compiler unexpectedly panicked. this is a bug.
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/panicking.rs:72:14

   2: core::panicking::assert_failed_inner
note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/panicking.rs:409:17

note: rustc 1.79.0 (129f3b996 2024-06-10) running on x86_64-unknown-linux-gnu
   3: core::panicking::assert_failed::<usize, usize>

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental=[REDACTED]
   4: rustc_incremental::persist::load::setup_dep_graph

   5: <rustc_interface::queries::Queries>::global_ctxt
note: some of the compiler flags provided by cargo are hidden
   6: rustc_interface::interface::run_compiler::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
```

</p>
</details>

**Comment by tgross35** [comment]
Does this issue persist after you do `cargo clean` or does it go away? Some things like this are cache-related.

> This issue only happens with the stable build

Meaning not on nightly?

**Comment by Kiwifuit** [comment]
> Does this issue persist after you do `cargo clean` or does it go away? Some things like this are cache-related.

It goes away

> Meaning the issue doesn't show up on nightly?

Yes, i have tried this command:
```
cargo +nightly check --workspace --message-format=json-diagnostic-rendered-ansi --manifest-path <Cargo.toml> --keep-going --all-targets
```

**Comment by tgross35** [comment]
> It goes away

Does it stay away, or is does running your command make it happen again?

**Comment by saethlin** [comment]
This ICE is almost certainly due to an incremental compilation file which was damaged by another process. This scenario seems to be so frequent that I've turned the ICE into a warning with a recovery path that just ignores the file. That change is in 1.80, which will be stable tomorrow.

**Comment by saethlin** [comment]
As far as I can tell, the ICE reported in this issue is caused by another process damaging rustc's build artifacts or by a buggy filesystem. That other process could be anything from an IDE running the compiler at the same time as a manual build to a find-and-replace that accidentally edits files in the target directory.

https://github.com/rust-lang/rust/pull/124686 contains a mitigation for this situation; the compiler now has a crude way to detect and ignore damaged incremental compilation files. You will still get a warning about the damaged file, but the build should proceed normally. That change shipped in Rust 1.80 which released today, so I am closing this issue and a few others which as far as I can tell all have the same root cause. You shouldn't see this ICE anymore, but please file a new issue if you see something like it on 1.80 or later.
